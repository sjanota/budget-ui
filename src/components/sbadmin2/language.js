import React, { createContext, useContext } from 'react';
import { useSBAdmin2 } from './context';
import PropTypes from 'prop-types';
import { capitalize } from '../../util/capitalize';

const DictionaryContext = createContext();

export function DictionaryProvider({ dictionaries, children }) {
  const { user } = useSBAdmin2();
  return (
    <DictionaryContext.Provider value={dictionaries[user.locale]}>
      {children}
    </DictionaryContext.Provider>
  );
}

DictionaryProvider.propTypes = {
  dictionaries: PropTypes.object,
  children: PropTypes.node,
};

export const useDictionary = () => useContext(DictionaryContext);

export function withDictionary(prop, Component, readPropName) {
  return props => {
    readPropName = readPropName || `read${capitalize(prop)}`;
    const readDict = props[readPropName];
    const newProps = { ...props };
    delete newProps[readPropName];
    const dictionary = useDictionary();
    if (readDict) {
      newProps[prop] = readDict(dictionary);
    }
    return <Component {...newProps} />;
  };
}
