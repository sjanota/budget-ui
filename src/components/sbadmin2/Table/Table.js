import React from 'react';
import PropTypes from 'prop-types';
import BootstrapTable from 'react-bootstrap-table-next';

import { withDictionary } from '../language';

function Table({ columns, columnNames, ...props }) {
  return (
    <BootstrapTable
      bootstrap4
      columns={addColumnNames(columns, columnNames)}
      {...props}
    />
  );
}

Table.propTypes = {
  columns: PropTypes.array.isRequired,
  columnNames: PropTypes.object.isRequired,
};

function addColumnNames(columns, dictionary) {
  return columns.map(c => ({ ...c, text: dictionary[c.dataField] || '' }));
}

export default withDictionary('columnNames', Table);
