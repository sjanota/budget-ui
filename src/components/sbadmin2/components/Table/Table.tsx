import React from 'react';
import BootstrapTable, {
  BootstrapTableProps,
  Column,
} from 'react-bootstrap-table-next';

import { withDictionary } from '../../language';

interface Dictionary {
  [key: string]: string;
}

interface Props<fieldIds extends string = string>
  extends BootstrapTableProps<fieldIds> {
  columnNames: Dictionary;
}

function Table({ columns, columnNames, ...props }: Props) {
  return (
    <BootstrapTable
      bootstrap4
      columns={addColumnNames(columns, columnNames)}
      {...props}
    />
  );
}

function addColumnNames<fieldIds extends string = string>(
  columns: Column<fieldIds>[],
  dictionary: Dictionary
) {
  return columns.map(c => ({ ...c, text: dictionary[c.dataField] || '' }));
}

export default withDictionary('columnNames', Table);
