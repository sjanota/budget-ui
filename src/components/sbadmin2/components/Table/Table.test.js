import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import BootstrapTable from 'react-bootstrap-table-next';
import Table from './Table';
import { DictionaryContext } from '../../language';

it('Table renders with minimal props', () => {
  const component = renderer.create(
    <Table
      columns={[{ dataField: 'id' }]}
      columnNames={{ id: 'ID' }}
      data={[{ id: 'id' }]}
      keyField="id"
    />
  );
  expect(component).toMatchSnapshot();
});

describe('Table', () => {
  const component = mount(
    <Table
      columns={[{ dataField: 'id' }]}
      columnNames={{ id: 'ID' }}
      data={[{ id: 'id' }]}
      keyField="id"
    />
  );

  it('adds column names', () => {
    expect(component.find(BootstrapTable)).toHaveProp('columns', [
      { dataField: 'id', text: 'ID' },
    ]);
  });

  it('creates table in bootstrap4 mode', () => {
    expect(component.find(BootstrapTable)).toHaveProp('bootstrap4', true);
  });

  it('can read column names from dictionary', () => {
    const component = mount(
      <DictionaryContext.Provider value={{ id: 'ID' }}>
        <Table
          columns={[{ dataField: 'id' }]}
          readColumnNames={d => d}
          data={[{ id: 'id' }]}
          keyField="id"
        />
      </DictionaryContext.Provider>
    );
    expect(component.find(BootstrapTable)).toHaveProp('columns', [
      { dataField: 'id', text: 'ID' },
    ]);
  });
});
