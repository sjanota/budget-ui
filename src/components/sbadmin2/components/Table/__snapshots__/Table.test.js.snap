// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders with minimal props 1`] = `
<div
  className="react-bootstrap-table"
>
  <table
    className="table table-bordered"
  >
    <thead>
      <tr>
        <th
          tabIndex={0}
        >
          ID
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className={null}
        style={Object {}}
      >
        <td>
          id
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
